docs/source/testing/mocking-responses.mdx

```diff
@@ -0,0 +1,181 @@
+---
+title: Mocking responses
+---
+
+Apollo Kotlin provides two tools to mock responses in tests:
+
+- `TestNetworkTransport`: allows to specify the GraphQL responses returned by `ApolloClient`
+- `MockServer`: a low-level HTTP server that can be used to mock any responses
+
+In most cases, `TestNetworkTransport` will be more straightforward to use in your tests, while `MockServer` can be
+useful to test specific server behavior, such as error cases, custom headers, timeouts, etc.
+
+## `TestNetworkTransport`
+
+This is a high level test API which allows to specify the responses that will be returned by your `ApolloClient`.
+
+Add the dependency to your project's `build.gradle` file:
+
+```kotlin:title=build.gradle[.kts]
+dependencies {
+  testImplementation("com.apollographql.apollo3:apollo-testing-support:x.y.z")
+}
+```
+
+Enable the `TestNetworkTransport` by passing it to the `ApolloClient` builder:
+
+```kotlin{2}
+val apolloClient = ApolloClient.Builder()
+    .networkTransport(TestNetworkTransport())
+    .build()
+```
+
+Either keep a reference to the `TestNetworkTransport` instance, or use the `testNetworkTransport` extension function to
+retrieve it:
+
+```kotlin
+val testNetworkTransport = apolloClient.testNetworkTransport
+```
+
+You can then use the `register` function to specify the responses that will be returned in function of the received
+operations:
+
+```kotlin{7}
+val query = GetHeroQuery("001")
+val testData = GetHeroQuery.Data {
+  hero = droidHero {
+    name = "R2D2"
+  }
+}
+apolloClient.testNetworkTransport.register(query, testData)
+
+val actual = apolloClient.query(query).execute().data!!
+assertEquals(testData.hero.name, actual.hero.name)
+```
+
+The `register` function accepts an `Operation` (here, the query) and the `ApolloResponse` to be returned when it is
+executed. As a shortcut you can pass a `Data` directly instead of an `ApolloResponse`, like in the example above.
+
+Note how you can use the [Test builders](./test-builders/) to easily create the data.
+
+### Advanced usage
+
+In some cases it can be useful to have different responses returned when executing the same operations consecutively.
+
+To that effect, instead of mapping operations to responses with `register`, it is possible to **enqueue** responses:
+they will be returned in the order they were added. To do so, pass a `QueueTestNetworkTransportHandler` to
+the `TestNetworkTransport`, and use its `enqueue` function:
```

Comment:
I'd keep the symmetry with MockServer advanced usage and make this section primarily about defining a custom `TestNetworkTransportHandler`

Reading that doc, I wonder if we should even push the symmetry further and have both `TestNetworkTransport` and `MockServer` use a queue by default. That will make it easier to learn one coming from the other. Or (thinking out loud) have both `Map` and `Queue` options by detault:

```kotlin
class TestNetworkTransport(handler: TestNetworkTransportHandler = QueueTestNetworkTransportHandler) {

  fun enqueue(response: ApolloResponse) // throws if handler is not QueueTestNetworkTransportHandler
  fun register(operation: Operation, response: ApolloResponse) // throws if handler is not RegistryTestNetworkTransportHandler
}

class MockServer(handler: MockServerHandler) {

  fun enqueue(response: HttpResponse)
  fun register(request: HttpRequest, response: HttpResponse)
}

```

Reply:
Good point!

On the MockServer side, I didn't make an equivalent `MapMockServerHandler` at all (not 100% sure it would be useful, and raises a few questions about how `MockRequest` equality should be implemented - I'll have one more look at it).

For `TestNetworkTransport`, I felt `MapTestNetworkTransportHandler` was a sensitive default but tbh I didn't think too much about it :). In the end it'd be nice if the default reflects on "how most users use it", but we don't really know, and I agree that having the same default as the MockServer is a plus (yay consistency!). I'll change the default to use Queue handler, and also add the `enqueue` method (and keep the register one), and update the doc.